lib/tbot/cli/cli.go

```diff
@@ -0,0 +1,225 @@
+	TokenEnvVar       = "TELEPORT_BOT_TOKEN"
+	ProxyServerEnvVar = "TELEPORT_PROXY"
+)
+
+var log = logutils.NewPackageLogger(teleport.ComponentKey, teleport.ComponentTBot)
+
+// CommandRunner defines a contract for `TryRun` that allows commands to
+// either execute (possibly returning an error), or pass execution to the next
+// command candidate.
+type CommandRunner interface {
+	TryRun(cmd string) (match bool, err error)
+}
+
+// MutatorAction is an action that is called by a config mutator-style command.
+type MutatorAction func(mutator ConfigMutator) error
+
+// genericMutatorHandler supplies a generic `TryRun` that works for all commands
+// that - broadly - load config, mutate that config, and run an action. It's
+// meant to be embedded within a command struct to provide the `TryRun`
+// implementation.
+type genericMutatorHandler struct {
+	cmd     *kingpin.CmdClause
+	mutator ConfigMutator
+	action  MutatorAction
+}
+
+// newGenericMutatorHandler creates a new generic genericMutatorHandler that
+// provides a generic `TryRun` implementation.
+func newGenericMutatorHandler(cmd *kingpin.CmdClause, mutator ConfigMutator, action MutatorAction) *genericMutatorHandler {
+	return &genericMutatorHandler{
+		cmd:     cmd,
+		mutator: mutator,
+		action:  action,
+	}
+}
+
+func (g *genericMutatorHandler) TryRun(cmd string) (match bool, err error) {
+	switch cmd {
+	case g.cmd.FullCommand():
+		err = g.action(g.mutator)
+	default:
+		return false, nil
+	}
+
+	return true, trace.Wrap(err)
+}
+
+// ConfigMutator is an interface that can apply changes to a BotConfig.
+type ConfigMutator interface {
+	ApplyConfig(cfg *config.BotConfig, l *slog.Logger) error
+}
+
+// genericExecutorHandler is a helper that can be embedded to provide a simpler
+// TryRun implementation that just runs a function. These functions can be
+// passed in while building the CLI to more easily glue behaviors together, or
+// specified directly.
+type genericExecutorHandler[T any] struct {
+	cmd  *kingpin.CmdClause
+	args *T
+
+	// actions is a list of functions to run when `TryRun` matches the `cmd`.
+	// Generally at most one action should be exposed to the top level glue in
+	// main, but commands might want to inject some handler logic for e.g.
+	// flag migrations.
+	actions []func(*T) error
+}
+
+// newGenericExecutorHandler creates a genericExecutorHandler with the given
+// command and action to execute when that command is matched.
+func newGenericExecutorHandler[T any](cmd *kingpin.CmdClause, args *T, actions ...func(*T) error) *genericExecutorHandler[T] {
+	return &genericExecutorHandler[T]{
+		cmd:     cmd,
+		args:    args,
+		actions: actions,
+	}
+}
+
+func (e *genericExecutorHandler[T]) TryRun(cmd string) (match bool, err error) {
+	switch cmd {
+	case e.cmd.FullCommand():
+		for _, action := range e.actions {
+			err = action(e.args)
+			if err != nil {
+				break
+			}
+		}
+	default:
+		return false, nil
+	}
+
+	return true, trace.Wrap(err)
+}
```

Comment:
TimB's on leave until next week - I figure that since this PR is already pretty large - the overall idea of extracting this to something reusable makes sense but we perhaps ought to do that in another PR. I think there's a few finishing touches/refinements we'll want to add if other folks are going to use this.